File: 25-reverse-nodes-in-k-group/reverse-nodes-in-k-group.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* kthNode(ListNode* head, int k) {
        ListNode* temp = head;
        int cnt = 1;
        while (temp != nullptr) {
            if (cnt == k)
                return temp;
            temp = temp -> next;
            cnt++;
        }
        return nullptr;
    }

    ListNode* reverseLL(ListNode* head) {
        ListNode* temp = head;
        ListNode* prev = nullptr;
        while (temp != nullptr) {
            ListNode* next = temp -> next;
            temp -> next = prev;
            prev = temp;
            temp = next;
        }
        return temp;
    }
    
    ListNode* reverseKGroup(ListNode* head, int k) {
        ListNode* temp = head;
        ListNode* nextNode = nullptr;
        ListNode* prevNode = nullptr;
        while(temp != nullptr){
            ListNode* kNode = kthNode(temp, k);
            if(kNode == nullptr){
                if(prevNode != nullptr)
                    prevNode -> next = temp;
                break;
            }
            nextNode = kNode -> next;
            kNode -> next = nullptr;
            reverseLL(temp);
            if(head == temp){
                head = kNode;
            }else{
                prevNode -> next = kNode;
            }
            prevNode = temp;
            temp = nextNode;
        }
        return head;
    }
};

File: 25-reverse-nodes-in-k-group/reverse-nodes-in-k-group.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} k
 * @return {ListNode}
 */

var reverseLinkedList = function (head) {
    let temp = head;
    let prev = null;
    while (temp) {
        let front = temp.next;
        temp.next = prev;
        prev = temp;
        temp = front;
    }
    return prev;
}

var getKthNode = function (temp, k) {
    k -= 1;
    while (temp && k > 0) {
        k--;
        temp = temp.next;
    }
    return temp;
}

var reverseKGroup = function (head, k) {
    let temp = head;
    let prevLast = null;
    while (temp) {
        let kThNode = getKthNode(temp, k);
        if (kThNode === null) {
            if (prevLast)
                prevLast.next = temp;
            break;
        }
        let nextNode = kThNode.next;
        kThNode.next = null;
        reverseLinkedList(temp);
        if (temp === head)
            head = kThNode;
        else
            prevLast.next = kThNode;
        prevLast = temp;
        temp = nextNode;
    }
    return head;
};